crates/astria-sequencer/src/mempool.rs

```diff
@@ -43,212 +55,236 @@ impl PartialOrd for TransactionPriority {
     }
 }
 
-impl TransactionPriority {
-    pub(crate) fn new(transaction_nonce: u32, current_account_nonce: u32) -> anyhow::Result<Self> {
-        if transaction_nonce < current_account_nonce {
+#[derive(Clone, Debug)]
+pub(crate) struct EnqueuedTransaction {
+    tx_hash: [u8; 32],
+    signed_tx: Arc<SignedTransaction>,
+}
+
+impl EnqueuedTransaction {
+    fn new(signed_tx: SignedTransaction) -> Self {
+        Self {
+            tx_hash: signed_tx.sha256_of_proto_encoding(),
+            signed_tx: Arc::new(signed_tx),
+        }
+    }
+
+    fn priority(&self, current_account_nonce: u32) -> anyhow::Result<TransactionPriority> {
+        let Some(nonce_diff) = self.signed_tx.nonce().checked_sub(current_account_nonce) else {
             return Err(anyhow::anyhow!(
-                "transaction nonce {} is less than current account nonce {}",
-                transaction_nonce,
-                current_account_nonce
+                "transaction nonce {} is less than current account nonce {current_account_nonce}",
+                self.signed_tx.nonce()
             ));
-        }
+        };
 
-        Ok(Self {
-            transaction_nonce,
-            current_account_nonce,
+        Ok(TransactionPriority {
+            nonce_diff,
         })
     }
-}
 
-/// [`BasicMempool`] is a simple mempool implementation that stores transactions in a priority queue
-/// ordered by nonce.
-///
-/// Future extensions to this mempool can include:
-/// - maximum mempool size
-/// - fee-based ordering
-/// - transaction expiration
-pub(crate) struct BasicMempool {
-    queue: DoublePriorityQueue<[u8; 32], TransactionPriority>,
-    hash_to_tx: HashMap<[u8; 32], SignedTransaction>,
-}
+    pub(crate) fn tx_hash(&self) -> [u8; 32] {
+        self.tx_hash
+    }
 
-impl BasicMempool {
-    #[must_use]
-    fn new() -> Self {
-        Self {
-            queue: DoublePriorityQueue::new(),
-            hash_to_tx: HashMap::new(),
-        }
+    pub(crate) fn signed_tx(&self) -> Arc<SignedTransaction> {
+        self.signed_tx.clone()
     }
 
-    #[must_use]
-    pub(crate) fn iter(&self) -> BasicMempoolIter {
-        BasicMempoolIter {
-            iter: self.queue.iter(),
-            hash_to_tx: &self.hash_to_tx,
-        }
+    pub(crate) fn address(&self) -> Address {
+        Address::from_verification_key(self.signed_tx.verification_key())
```

Comment:
we should probably just make this a method on `SignedTransaction`, would be useful elsewhere

Reply:
reviewed this before #1111 so i think it's fine with that :)